### ISSUE TITLE
null/undefined initial atom value confuses typescript

### ISSUE THREAD

**Issue by ahfarmer** [pmndrs/jotai#550]
Whenever I set null or undefined as the initial value on an atom, TypeScript infers the type as `Atom` instead of `WritableAtom`.

Working case: initial value is `false` so type is shown as `WritableAtom`:
![Screen Shot 2021-06-23 at 3 12 58 PM](https://user-images.githubusercontent.com/597825/123174928-c6499b80-d435-11eb-8a32-4b8105610678.png)

Non-working case: initial value is `null` so type is just `Atom`:

![Screen Shot 2021-06-23 at 3 13 22 PM](https://user-images.githubusercontent.com/597825/123175008-e8dbb480-d435-11eb-85c0-093cefc3c5ef.png)

This shows up as a type error on `useUpdateAtom`:
![Screen Shot 2021-06-23 at 3 13 15 PM](https://user-images.githubusercontent.com/597825/123175047-f729d080-d435-11eb-8d11-32eedb9cf05c.png)

My current workaround: caste to `PrimitiveAtom`:
```typescript
const myAtomConfig = atom<boolean | null>(null) as PrimitiveAtom<
  boolean | null
>;
```

**Comment by dai-shi** [comment]
Oh... Nice catch. Do you happen to know how to fix??

**Comment by ahfarmer** [comment]
Looks like TypeScript is choosing the wrong overload for atom(). It is choosing between these two:

```typescript
export declare function atom<Value>(read: Read<Value>): Atom<Value>;
export declare function atom<Value>(initialValue: Value): [Value] extends [Function] ? never : PrimitiveAtom<Value> & WithInitialValue<Value>;
```

When we have a non-null like `atom("hi")` it can see that "hi" is not a function and chooses the correct overload (the second one).

But when we have `atom(null)` it matches the first overload (the one for read-only derived atoms) and uses that because it appears first.

I couldn't figure out how to tell TypeScript that `null` does not qualify as `Read<Value>`, but I can fix it by adding an additional overload specifically for `null | undefined`:

```typescript
export declare function atom<Value>(initialValue: null | undefined): [Value] extends [Function] ? never : PrimitiveAtom<Value> & WithInitialValue<Value>;
export declare function atom<Value>(read: Read<Value>): Atom<Value>;
export declare function atom<Value>(initialValue: Value): [Value] extends [Function] ? never : PrimitiveAtom<Value> & WithInitialValue<Value>;
```

A little verbose but it works. I sent a PR.

**Comment by dai-shi** [comment]
Hmm, I can't reproduce this.

Oh, I guess your strictNullChecks is off?
https://www.typescriptlang.org/tsconfig#strictNullChecks

It would be not possible without "strict: true".

**Comment by ahfarmer** [comment]
Ah, yes you are correct. This issue only occurs with `strictNullChecks=false`.

**Comment by dai-shi** [comment]
Is there any reason you can't enable `strictNullChecks`. I don't think we can solve it with `strictNullChecks=false`

#554 isn't necessary with `strictNullChecks=true`. And, with `strictNullChecks=false` #554 is not enough (it doesn't infer correctly, and other tests fail).

**Comment by ahfarmer** [comment]
I do plan to enable `strictNullChecks` soon but it will be a pretty large effort in my project - hundreds of errors to fix when I turn it on.

`strictNullChecks=false` is the default with a brand new `tsconfig.json` from `tsc --init` so if this goes unfixed, Jotai might seem broken to new users. The only reason I have `strictNullChecks=false` is because it is the default and I never changed it.

I also have about 10 dependencies that I use types for, and I'm not getting any errors from the other libraries, so I believe this should be possible. Unless Jotai is doing something totally unique here? Perhaps it is doing more overloading than most other libraries?

In any case I'll see if I can convince you with a solution that is better than #544 - but probably not until next week.

**Comment by ahfarmer** [comment]
I double checked and I was not correct. `tsc --init` actually creates a file that does have `strict: true`. However, if you have an empty `tsconfig.json`, it will give you the default which is `strict: false`. ( https://www.typescriptlang.org/tsconfig#strict )

I ended up changing my project to `strict: true` so this issue isn't important to me anymore - feel free to close.

**Comment by dai-shi** [comment]
> the default which is strict: false

Yes, I know. This is very unfortunate and annoying.

Closing. Look forward to another issue report!

**Comment by alexluong** [comment]
Hi there, is there any way we can re-open this?

I'm hoping we can use `jotai` but we have `strict: false` because we're using GraphQL codegen and it's a little tricky to be strict with it.

**Comment by dai-shi** [comment]
It’s closed because I think it’s impossible to fix in a reasonable way. But I might be wrong. Happy to consider suggestions from anyone. 
The workaround for you should be explicit typing. How would that work for you?

**Comment by alexluong** [comment]
> The workaround for you should be explicit typing.

Can you give an example of what you mean?

**Comment by alexluong** [comment]
Just FYI I ended up setting `struct: true` as well, but I think it would be good to know the workaround if possible! Thanks for being very responsive and I really enjoy using `jotai` so far!

**Comment by dai-shi** [comment]
> > The workaround for you should be explicit typing.
> 
> Can you give an example of what you mean?

Sorry, it's not explicit typing, but type assertion.
Like OP noted:
```js
const myAtomConfig = atom<boolean | null>(null) as PrimitiveAtom<
  boolean | null
>;
```
